frontend/src/bundles/common/components/calendar/calendar.tsx

```diff
@@ -0,0 +1,186 @@
+import clsx from 'clsx';
+import { type ChangeEvent,useCallback } from 'react';
+import { useEffect, useRef, useState } from 'react';
+import { Calendar as ReactCalendar } from 'react-calendar';
+import { type Control, type FieldPath, type FieldValues } from 'react-hook-form';
+
+import ArrowImage from '~/assets/img/arrow.svg';
+
+import { useAppForm, useFormController } from '../../hooks/hooks';
+import { type CalendarDate, type CalendarMonth } from '../../types/types';
+import { Toggle } from '../components';
+import { CalendarMonths, DEFAULT_DATE_PAYLOAD, monthRegex,yearRegex } from './constants/calendar.constants';
+import styles from './styles.module.scss';
+
+type Properties<T extends FieldValues> = {
+    control: Control<T, null>;
+    name: FieldPath<T>;
+    showPresent?: boolean;
+    className?: string
+};
+
+const Calendar = <T extends FieldValues>({
+    showPresent = false,
+    className = '',
+    name,
+    control
+}: Properties<T>): JSX.Element => {
+    const { control: innerControl } = useAppForm<CalendarDate>({
+        defaultValues: DEFAULT_DATE_PAYLOAD,
+    });
+
+    const { field } = useFormController({ name, control });
+
+    const { field: presentField } = useFormController({ name: 'present', control: innerControl });
+
+    const reference = useRef<HTMLDivElement>(null);
+
+    const [text, setText] = useState('');
+
+    const handleTextChange = useCallback((error: ChangeEvent<HTMLInputElement>): void => {
+        setText(error.target.value);
+
+        const year = error.target.value.match(yearRegex);
+        if (year) {
+            setYear(Number(year[0]));
+        }
+
+        const monthMatch = error.target.value.match(monthRegex);
+
+        if (monthMatch) {
+            for(const month of CalendarMonths) {
+                if (
+                    monthMatch[0]
+                        .toLowerCase()
+                        .includes(month.name.toLowerCase()) ||
+                    month.name
+                        .toLowerCase()
+                        .startsWith(monthMatch[0].toLowerCase())
+                ) {
+                    setMonth(month);
+                    setSelected(month.num);
+                }
+            }
+        } else {
+            setMonth(null);
+            setSelected(0);
+        }
+    }, []);
+
+    const [focused, setFocused] = useState(false);
+
+    const handleInputFocus = useCallback((focused: boolean) => 
+        clsx(styles.calendar__date_input, styles.focused, focused && styles.calendar__date_input)
+    , []);
+
+    const setCurrentlyFocused = useCallback((): void => setFocused(true), []);
+    const setUnfocused = useCallback((): void => setFocused(false), []);
+
+    const [selected, setSelected] = useState<number | null>(null);
+
+    const [year, setYear] = useState(new Date().getFullYear());
+    const [month, setMonth] = useState<CalendarMonth | null>(null);
```

Comment:
Move `useState` to the top of the component

Reply:
done